src/x509_parser.rs

```diff
@@ -247,35 +247,63 @@ fn parse_extensions(i:&[u8]) -> IResult<&[u8],Vec<X509Extension>,BerError> {
 }
 
 
-fn parse_tbs_certificate(i:&[u8]) -> IResult<&[u8],TbsCertificate,BerError> {
-    parse_der_struct!(
-        i,
-        TAG DerTag::Sequence,
-        version:     parse_version >>
-        serial:      map_opt!(parse_der_integer, |x:DerObject| x.as_biguint()) >>
-        signature:   parse_algorithm_identifier >>
-        issuer:      parse_name >>
-        validity:    parse_validity >>
-        subject:     parse_name >>
-        subject_pki: parse_subject_public_key_info >>
-        issuer_uid:  parse_issuer_unique_id >>
-        subject_uid: parse_subject_unique_id >>
-        extensions:  parse_extensions >>
-        (
-            TbsCertificate{
-                version,
-                serial,
-                signature,
-                issuer,
-                validity,
-                subject,
-                subject_pki,
-                issuer_uid,
-                subject_uid,
-                extensions
+fn parse_tbs_certificate(i: &[u8]) -> IResult<&[u8], TbsCertificate, BerError> {
+    use nom::{Offset, Slice, error::ParseError, combinator::map};
+    // A combinator which returns the consumed input and the result.
+    // It is similar to the recognize combinator from nom.
+    fn keep_input<'a, O, E:ParseError<&'a [u8]>, F>(parser:F) -> impl Fn(&'a [u8]) -> IResult<&'a [u8], (&'a [u8], O), E>
```

Comment:
Hum, this seems overly complicated. I like combinators, but I'd clearly prefer simplifying code using something like (in the main function, no need for closures/clones)
```rust
let start = i;
let (rem,mut tbs) = parse_tbs(i)?;
let len = start.offset(rem);
// store i[..len] in tbs
Ok((rem, tbs))
```

This looks much more readable